### PR TITLE
Track Runme runs (logged into Stateful-only)

### PR DESCRIPTION
**Pull request by peraltafederico**
- Track runme cell run events and save it to the Cloud database only when the user is logged in.
- Some type narrowing clean up to avoid casting

Depends on https://github.com/stateful/platform/pull/1213

### CODE REVIEW

File: src/extension/kernel.ts [hunk 1]
```diff
@@ -832,7 +901,11 @@ export class Kernel implements Disposable {
       'cell.success': successfulCellExecution?.toString(),
       'cell.mimeType': annotations.mimeType,
     })
-    runmeExec.end(successfulCellExecution, Date.now())
+    const endTime = Date.now()
+
+    this.#handleTrackCellRun(cell, successfulCellExecution, startTime, endTime)
```
Comment (pastuxso) [inline src/extension/kernel.ts:906]: @peraltafederico I think the call should be conditioned to the telemetry flag `isTelemetryEnabled`. Take a look at https://github.com/stateful/vscode-runme/commit/4c11a9c2dbaf8a56e9dcac07febbc5d2f7eae744#diff-2b0a484742f2baf7a6b33fec5aa69dfbf17fbe089363cec3549d0b88f664b1b4R23

File: src/extension/kernel.ts [hunk 2]
```diff
@@ -734,6 +738,70 @@ export class Kernel implements Disposable {
...
 
+  async #handleTrackCellRun(
+    cell: NotebookCell,
+    successfulCellExecution: boolean,
+    startTime: number,
+    endTime: number,
+  ): Promise<void> {
+    const session = await getPlatformAuthSession(false)
+
+    if (!this.serializer?.isLifecycleIdentity(RunmeIdentity.ALL || !session)) {
```
Comment (pastuxso) [inline src/extension/kernel.ts:749]: I think the `||` should after closing the call `isLifecycleIdentity`.

```suggestion
    if (!this.serializer?.isLifecycleIdentity(RunmeIdentity.ALL) || !session) {
```

Comment (pastuxso) [reply src/extension/kernel.ts:749]: Apart of that, I think is not necessary to add a new method to the serializer. It's enough calling the runtime setting value directly from `ContextState.getKey<ServerLifecycleIdentity>(NOTEBOOK_LIFECYCLE_ID)`.

**Review by pastuxso — CHANGES_REQUESTED** [review]

File: src/extension/kernel.ts (hunk 2 above)
Comment (peraltafederico) [reply src/extension/kernel.ts:749]: Ha! Totally. Last minute change to make readability better and I missed that nonsense parenthesis. Good catch!

Comment (peraltafederico) [reply src/extension/kernel.ts]: Fixed in 99e68515e54b2cb2745ea9658518dfc16b5cd32e

File: src/extension/kernel.ts (hunk 1 above)
Comment (peraltafederico) [reply src/extension/kernel.ts:914]: Fixed in 99e68515e54b2cb2745ea9658518dfc16b5cd32e

**Comment by sourishkrout** [comment]
I'll look at it over the weekend.

**Review by sourishkrout — APPROVED** [review]
I want to detach rendering/packaging and delivering the message for the former to be sync and the latter to be async. Merging for now.

I will refactor this to be sync+async next week.

**Pull request merged**